### ISSUE TITLE
"Connection closed normally" WebSocket error

### ISSUE THREAD

**Issue by Tangent128** [seanmonstar/warp#593]
When I use a WebSocket as a Stream, a regularly-closed-by-the-client connection will end with a ConnectionClosed "error" from tungstenite; this creates noise in my error logs suggesting that something has gone wrong when it hasn't.

It would be convenient if that error ended the Stream instead of being exposed as an opaque WarpError; it appears that [used to be the case](https://github.com/seanmonstar/warp/commit/9dacd7d9977e7894aada7ea4cf71754aa41e1033#diff-9119b6774ba273de868213a966affe87L292), but I couldn't determine if it was intentionally changed.

**Comment by seanmonstar** [comment]
I don't think it was intentional. I agree that a clean close shouldn't be an error, when we already have `None` to represent that.

**Comment by jxs** [comment]
@Tangent128 does #596 fix the issue?

**Comment by Tangent128** [comment]
Thanks for the quick response; it is with egg on my face that I have to say testing the fix out showed that I misdiagnosed the issue I was seeing, however;

As it turns out, tokio-tungstenite appears to [already do the mapping to None](https://github.com/snapview/tokio-tungstenite/blob/master/src/lib.rs#L274), so using the websocket Stream already works as expected. (#596 won't hurt anything, it just turns out that case shouldn't actually happen after all)

The issues I was seeing came from the *Sink* end; my handler was joining two async blocks for the source and sink (and thus their errors), but wasn't dropping the sink block when the source ended. Wrapping it in an Abortable did fix my issue.

This issue can be closed; a more ergonomic way to tell "this socket was closed elsewhere, you just didn't notice yet" would be nice, but I can't really see how that would fit into the existing interface.
